### PR TITLE
[Log Collector] Improve memory consumption

### PR DESCRIPTION
**Pull request by TomerShor**
Improve log collector memory consumption:

- Implement a sized byte buffer pool, which is blocking if the pool has allocated the configured max size.
- Replace `CopyBuffer` with the non-blocking `stream.Read`, and handle the errors according to num bytes read
- When streaming pod logs, keep the log file open in append mode instead of constantly opening and closing the file.
- Throttle warning of failure to find run pods, as sometimes pods take time to start/create/etc.

### CODE REVIEW

File: go/pkg/common/bufferpool/sizedbytepool.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+// Put returns a byte buffer to the pool.
+// If the buffer is too small, it is discarded.
+func (p *SizedBytePool) Put(buf []byte) {
+	if cap(buf) < p.bufferSize {
+		return
+	}
+	select {
+	case p.bufferChan <- buf[:p.bufferSize]:
+	default:
+	}
```
Comment (quaark) [inline go/pkg/common/bufferpool/sizedbytepool.go:61]: Add a comment explaining the magic happening here :)

File: go/pkg/common/bufferpool/bufferpool_test.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	suite.Require().NoError(err, "Failed to create logger")
+
+	suite.logger.Info("Setup complete")
+}
+
+func (suite *sizedBytePoolTestSuite) TearDownSuite() {
+	suite.logger.Info("Tear down complete")
+}
+
+func (suite *sizedBytePoolTestSuite) TestByteBufferPoolBlocking() {
```
Comment (quaark) [inline go/pkg/common/bufferpool/bufferpool_test.go:44]: great test! 🚀

File: go/pkg/services/logcollector/server.go [hunk 3]
```diff
@@ -418,48 +419,60 @@ func (s *Server) startLogStreaming(ctx context.Context,
...
-				"runUID", runUID,
-				"err", streamErr)
+
+			// first 3 errors are not logged - they are expected if pod is not ready yet
+			if streamErrCount > 3 {
+				s.Logger.WarnWithCtx(ctx,
+					"Failed to get pod log stream, retrying",
+					"runUID", runUID,
+					"err", streamErr)
+			}
```
Comment (quaark) [inline go/pkg/services/logcollector/server.go:456]: If you were preventing log spam - ignore me.
If not, maybe having the err log but debug (so less spammy) and expected would be nice
```suggestion
			if streamErrCount > 3 {
				s.Logger.WarnWithCtx(ctx,
					"Failed to get pod log stream, retrying",
					"runUID", runUID,
					"err", streamErr)
			} else {
				s.Logger.DebugWithCtx(ctx,
					"Failed to get pod log stream as pod is still starting up, retrying",
					"runUID", runUID,
					"err", streamErr)
			}
```

File: go/pkg/services/logcollector/server.go [hunk 4]
```diff
@@ -477,71 +490,47 @@ func (s *Server) startLogStreaming(ctx context.Context,
...
+		if err != nil {
+			s.Logger.WarnWithCtx(ctx, "Failed to read pod log",
+				"err", err.Error(),
+				"runUID", runUID)
+			return false, errors.Wrap(err, "Failed to read pod logs")
+		}
+
+		// nothing happened, continue
+		return true, nil
 	}
```
Comment (quaark) [inline go/pkg/services/logcollector/server.go:534]: You are already returning `true, nil` in the `//sanity` after this if. So you can consolidate these 2 into:
```
if err != nil && numBytesRead == 0  {
	s.Logger.WarnWithCtx(ctx, "Failed to read pod log",
 		"err", err.Error(),
 		"runUID", runUID)
 	return false, errors.Wrap(err, "Failed to read pod logs")
}

**Review by quaark — CHANGES_REQUESTED** [review]
Great stuff. A couple of minor comments.

File: go/pkg/common/bufferpool/sizedbytepool.go (hunk 1 above)
Comment (TomerShor) [reply go/pkg/common/bufferpool/sizedbytepool.go:61]: Simplified it..

File: go/pkg/services/logcollector/server.go (hunk 3 above)
Comment (TomerShor) [reply go/pkg/services/logcollector/server.go:456]: Exactly prevented log spam, as I wrote above it:
```
first 3 errors are not logged - they are expected if pod is not ready yet
```

**Review by quaark — APPROVED** [review]
LGTM

**Pull request merged**